### ISSUE TITLE
Unable to send the same entry twice in the same block

### ISSUE THREAD

**Issue by schalk-b** [FactomProject/factomd#503]
Hello

My understanding is that the protocol supports sending the same entry multiple times - which is also mentioned on https://www.factom.com/devs/docs/guide/factom-data-structures where it has an example of duplicate entries in the same block.

However, I'm unable to get this working within the same block, it works fine if it gets sent to different blocks.

**Reproduce:**
Commit an entry
Reveal the entry
Wait on Minute
Commit the same entry (but with a different timestamp so that the commit message is different to the first time)
Reveal the entry

**Result:**
All successful responses from the calls, however only the first commit/reveal makes it into the entry block. 

When checking the ack of the second commit by txid, I get the response:
{
    "jsonrpc": "2.0",
    "id": 0,
    "result": {
        "committxid": "5931cca67e959bb57b37a183adf1d4c0ba8c6730170a70ae238f9c9f323a3c1b",
        "entryhash": "",
        "commitdata": {
            "status": "Unknown"
        },
        "entrydata": {
            "status": ""
        }
    }
}

Also to add to this, once broadcasting the same entries, an entry got stuck in "NotConfirmed" when checking the pending-entries and this persisted for multiple blocks which prevented that entry from getting in (even if it was sent again).

**Comment by carryforward** [comment]
good question @schalk-b This was the case in M1, as things were easer on a single server.

It turned out to be a matter of a few things causing the behavior you are observing, and had adapted as the protocol has developed.

1. There must be duplicates allowed, because otherwise, the servers would have to track at a quick pace all entries that have ever happened.  Over time this would be a very large list to check against, with little benefit to the users.  This means that the application level needs to deal with potential duplicates.

2. almost all reasonable real world applications in the system won't be putting duplicate data into the system.  Not if they want to protect themselves against replay attacks.  

Any near-do-well can take an old entry you made, pay for it again with their ECs (months later), and have it put back in the chain it came from.  If your application does not take this into account, it can get disrupted by an outsider.

This means that if someone is updating a document process it needs to have a timestamp next to the document hash, or a serial number, or something like that.  This timestamp in the entry would be under a signature, so an outsider cannot change the content of your data without breaking the signature, causing your application to ignore it.

Only very odd things would have identical entries put into the same chain in quick succession.  They can be made non-identical by adding a random number into the mix, making the entries non-duplicate.

3. The reality of the system as we developed m2 was that tracking duplicates asynchronously across multiple p2p nodes was tricky for several reasons.  It made things a whole lot easier to only allow one identical entry over a short period of time.  Factoid and EC transaction are only allowable over the +- 1 hour timeframe that they are timestamped with. It made sense to disallow paying for duplicate entries, in addition to disallowing the duplicate ECs themselves.  By only allowing one EC transaction per hour per entry, that limits the ability to make duplicate entries to 1 per hour or so.  

There are edge cases with paying too few entry credits, which was an attack vector that was closed, but we don't need to get into that.

in conclusion, your application needs to worry about duplicates in the chain, and there may be edge cases during faulting where there might be duplicate entries in the same block, but you won't be able to make duplicate entries on a short term basis normally.

What problem are you trying to solve by using duplicate entries?

**Comment by schalk-b** [comment]
Hi Brian!

Sorry, I had gotten side tracked and forgotten about this issue that I raised. Thanks for your detailed explanation. 

I agree that this probably wouldn't be a high priority to resolve and more of an edge case.

I've been doing development on an API similar to Apollo which allows third party developers to create entries into factom and get callbacks when the entries hit certain states, like DBlockConfirmed. But I am now leaning towards preventing the ability to send the same entry twice and would rather just document in the API that duplicate entries can open up the possibility to replay attacks.

**Comment by PaulBernier** [comment]
Hijacking this thread. I didn't know it was even possible to submit twice the same entry in the same chain! Probably because I never waited 1-2h before trying to re submit. I haven't thought about it too much, but doesn't that lead some potential problems to have 2 entries with the exact same entryHash? I have always considered that an entryHash uniquely identify an entry, but I was wrong... so to identify an entry you need entryHash + entry block is was part of?

**Comment by drobertbaker** [comment]
Yes, if they are duplicates.
There are legitimate use cases for this (public signalling) and we felt we needed to provide this facility.
As pointed out above, if you want to use this approach and be able to tell your identical entries apart, you should employ timestamping.
